src/util/codec/convert.rs

```diff
@@ -133,13 +132,84 @@ fn get_valid_float_prefix(s: &str) -> &str {
         }
     }
 
-    if valid_len == 0 { "0" } else { &s[..valid_len] }
+    try!(handle_truncate(ctx, valid_len < s.len()));
+    if valid_len == 0 {
+        Ok("0")
+    } else {
+        Ok(&s[..valid_len])
+    }
+}
+
+/// It converts a valid float string into valid integer string which can be
+/// parsed by `i64::from_str`, we can't parse float first then convert it to string
+/// because precision will be lost.
+fn float_str_to_int_string(valid_float: Cow<str>) -> Result<Cow<str>> {
+    let mut dot_idx = None;
+    let mut e_idx = None;
+    for (i, c) in valid_float.chars().enumerate() {
+        match c {
+            '.' => dot_idx = Some(i),
+            'e' | 'E' => e_idx = Some(i),
+            _ => (),
+        }
+    }
+
+    if dot_idx.is_none() && e_idx.is_none() {
+        return Ok(valid_float);
+    }
+
+    let mut int_cnt;
+    let mut digits = String::with_capacity(valid_float.len());
```

Comment:
`digits` is unnecessary. We can use `int_cnt` and `digit_cnt` to detect if the int string is just a prefix of `valid_float` or has be be created from scratch.